crates/redis/src/lib.rs

```diff
@@ -52,30 +56,78 @@ impl TriggerExecutor for RedisTrigger {
     type RunConfig = NoArgs;
 
     async fn new(engine: TriggerAppEngine<Self>) -> Result<Self> {
-        let address = engine
+        let default_address: String = engine
             .trigger_metadata::<TriggerMetadata>()?
             .unwrap_or_default()
             .address;
 
-        let mut channel_components: HashMap<String, Vec<String>> = HashMap::new();
+        let mut addresses = HashSet::new();
+
+        let mut server_channels: HashMap<String, ChannelComponents> = HashMap::new();
 
         for (_, config) in engine.trigger_configs() {
-            channel_components
+            let address = config.address.clone().unwrap_or(default_address.clone());
+            addresses.insert(address.clone());
+            let server = server_channels.entry(address).or_default();
+            server
                 .entry(config.channel.clone())
                 .or_default()
                 .push(config.component.clone());
         }
         Ok(Self {
-            engine,
-            address,
-            channel_components,
+            engine: Arc::new(engine),
+            addresses: addresses.into_iter().collect(),
+            server_channels,
         })
     }
 
     /// Run the Redis trigger indefinitely.
     async fn run(self, _config: Self::RunConfig) -> Result<()> {
-        let address = &self.address;
+        let addresses = self.addresses.clone();
+        let tasks: Vec<_> = addresses
+            .into_iter()
+            .map(|address| {
+                let trigger = self.clone();
+                tokio::spawn(async move { trigger.run_listener(address).await })
+            })
+            .collect();
+
+        // wait for the first handle to be returned and drop the rest
+        let (_, _, rest) = futures::future::select_all(tasks).await;
+        drop(rest);
+        println!("exiting");
+        Ok(())
+    }
+}
+
+impl RedisTrigger {
+    // Handle the message.
+    async fn handle(&self, address: String, msg: redis::Msg) -> Result<()> {
+        let channel = msg.get_channel_name();
+        tracing::info!("Received message on channel {:?}", channel);
+
+        if let Some(channels) = self.server_channels.get(&address) {
```

Comment:
This shouldn't change on each message - there is no need to re-look-up the address each time.

There's some things about the code here that make me wonder if there's a listener type (one instance per server, holding that server's channel map, running that server's loop, essentially pulling out `run_listener` to hold its own state) trying to get out.  That may be more of a change than you fancy getting into though!

Reply:
I am not sure I follow. Are you suggesting that each listener only get the only the hashmap (channel, components) instead of the double hashmap that it currently gets?

Reply:
Yes.